Upload javadocs to gh-pages

abesto opened the pull request:
Creates a directory structure that looks like this:

```
javadoc-builddir/
├── index.html
└── 1.6.1-SNAPSHOT
    ├── index.html
    ├── zipkin
    │   ├── META-INF
    │   │   └── MANIFEST.MF
    │   ├── allclasses-frame.html
    │   ├── allclasses-noframe.html
    │   ├── constant-values.html
    │   ├── deprecated-list.html
...
```

And makes the contents available via GitHub Pages. The URL is, somewhat counter-intuitively, http://zipkin.io/zipkin/ (where `zipkin.io` is the domain used by all GitHub Pages sites of the `openzipkin` org, and `/zipkin` is the name of this repo).

Uploads javadocs only when NOT on a pull request build.

Did some testing, and tried to avoid problems, but we might still hit problems around `git checkout`, `git commit`, and uncommitted files. If all else fails, we can always use a separate  clone in, say, `/tmp`.

(Also did some research; there doesn't seem to exist a Maven plugin for doing this.)

abesto committed:
Upload javadocs to gh-pages

abesto commented:
Implementing (the simpler) half of https://github.com/openzipkin/zipkin/issues/1221. CC @adriancole

codefromthecrypt commented on travis/publish.sh:
```diff
@@ -122,11 +159,13 @@ elif is_travis_branch_master; then
 elif build_started_by_tag; then
   safe_checkout_master
   ./mvnw --batch-mode -s ./.settings.xml -Prelease -nsu -DreleaseVersion="$(release_version)" -Darguments="-DskipTests" release:prepare
+  javadoc_to_gh_pages
```
ah ok so this is for publishing against snapshots..

codefromthecrypt commented:
build failure unrelated.. rebase would fix it

codefromthecrypt commented:
maybe update RELEASE.md? Otherwise LGTM. Thanks for the help!

codefromthecrypt commented:
Also, probably putting a javadoc link reference to http://zipkin.io/zipkin/ at the bottom of the README with the other things.

abesto commented:
Rebased, added notes to `{RELEASE,README}.md`.

Happy to halp! In other news: just realized I can probably retroactively generate the javadocs for all previous releases, will see to that now.

codefromthecrypt commented:
Thanks tons!

abesto commented:
Done with the time traveling, http://zipkin.io/zipkin/ now has tons of yummy content. Built with https://github.com/openzipkin/zipkin/blob/gh-pages/rebuild-all-javadocs.sh

codefromthecrypt commented:
hit a snag on a release tag:

https://travis-ci.org/openzipkin/zipkin/builds/153257593

```
+javadoc_to_gh_pages
++print_project_version
++./mvnw help:evaluate -N -Dexpression=project.version
++grep -v '\['
Java HotSpot(TM) 64-Bit Server VM warning: ignoring option MaxPermSize=192m; support was removed in 8.0
+version=1.7.1-SNAPSHOT
+rm -rf javadoc-builddir
+builddir=javadoc-builddir/1.7.1-SNAPSHOT
++find . -name '*1.7.1-SNAPSHOT-javadoc.jar'
+git checkout gh-pages
error: pathspec 'gh-pages' did not match any file(s) known to git.
The command "./travis/publish.sh" exited with 1.
store build cache
```

codefromthecrypt commented:
ps thanks for making the step last! wise

codefromthecrypt commented:
little red here, too.. https://travis-ci.org/openzipkin/zipkin/builds/153262205

abesto commented:
Oops. I was wondering why I didn't see any new entries. Checking.

abesto commented:
So! The problem is that the script finds and extracts all the javadoc jars of the right version, but it looks like we have two of some of them, for instance:
- ./zipkin-server/target/local-repo/io/zipkin/java/zipkin-autoconfigure-ui/1.7.0/zipkin-autoconfigure-ui-1.7.0-javadoc.jar
- ./zipkin-autoconfigure/ui/target/zipkin-autoconfigure-ui-1.7.0-javadoc.jar -d javadoc-builddir/1.7.0/zipkin-autoconfigure-ui

More importantly: is it safe to just ignore all but the first jar per module?

codefromthecrypt commented:
local-repo is for integration testing the jar you just made. so, you
can safely ignore that path.
